apps/DisplayCluster/WallApplication.cpp

```diff
@@ -163,21 +167,40 @@ void WallApplication::startRendering()
     // Must be a queued connection to avoid infinite recursion.
     connect( this, SIGNAL( frameFinished( )),
              this, SLOT( renderFrame( )), Qt::QueuedConnection );
+
+    // swap sync; afterRendering signal is emitted before swapBuffers
+    connect( _renderContext->getWindow().get(), &WallWindow::afterRendering,
+             [this]()
+    {
+        _renderContext->getWindow()->openglContext()->functions()->glFinish();
+        if( _renderedFrames == 0 )
+            _wallChannel->globalBarrier();
+        ++_renderedFrames;
+    });
+
+    // trigger new renderloop after rendering & swap or quit application
+    connect( _renderContext->getWindow().get(), &WallWindow::frameSwapped,
+             [this]()
+    {
+        _renderContext->getWindow()->openglContext()->functions()->glFlush();
+        // expose event also causes swap, but don't trigger new frames in that
+        // case to not messup/deadlock swap- and object-sync
+        if( _renderedFrames == 1 )
```

Comment:
Then the if would check >=1, but the exact check for 1 matters. Only trigger new frames once per swap/render cycle. This fights against expose and and what not. I tried bool as well but failed miserably.